cli_tools/osconfig_agent/inventory.go

```diff
@@ -0,0 +1,138 @@
+//  Copyright 2017 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+// gce_inventory_agent gathers and writes instance inventory to guest attributes.
+package main
+
+import (
+	"bytes"
+	"compress/gzip"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"os"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/GoogleCloudPlatform/compute-image-tools/go/osinfo"
+	"github.com/GoogleCloudPlatform/compute-image-tools/go/packages"
+)
+
+const (
+	inventoryURL = reportURL + "/guestInventory"
+)
+
+type instanceInventory struct {
+	Hostname          string
+	LongName          string
+	ShortName         string
+	Version           string
+	Architecture      string
+	KernelVersion     string
+	InstalledPackages packages.Packages
+	PackageUpdates    packages.Packages
+	Errors            []string
+}
+
+func postAttributeCompressed(url string, body interface{}) error {
+	buf := &bytes.Buffer{}
+	b := base64.NewEncoder(base64.StdEncoding, buf)
+	zw := gzip.NewWriter(b)
+	w := json.NewEncoder(zw)
+	if err := w.Encode(body); err != nil {
+		return err
+	}
+
+	if err := zw.Close(); err != nil {
+		return err
+	}
+	if err := b.Close(); err != nil {
+		return err
+	}
+
+	return postAttribute(url, buf)
+}
+
+func writeInventory(state *instanceInventory, url string) {
```

Comment:
I wonder when will writeInventory be called with an url other than the constant inventoryURL and whether we need this as an input param?

Reply:
For tests we will want to change it (set it to the test server), tests that are currently unimplemented :)